redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/UIViewBox.kt

```diff
@@ -67,45 +72,62 @@ internal class UIViewBox : Box<UIView> {
         right = margin.end.toPx(),
       )
     }
-    value.setNeedsLayout()
+    value.invalidateSize()
   }
 
   override fun horizontalAlignment(horizontalAlignment: CrossAxisAlignment) {
     value.horizontalAlignment = horizontalAlignment
-    value.setNeedsLayout()
+    value.invalidateSize()
   }
 
   override fun verticalAlignment(verticalAlignment: CrossAxisAlignment) {
     value.verticalAlignment = verticalAlignment
-    value.setNeedsLayout()
+    value.invalidateSize()
   }
 
   internal class View : UIView(CGRectZero.readValue()) {
     var widthConstraint = Constraint.Wrap
     var heightConstraint = Constraint.Wrap
     var horizontalAlignment = CrossAxisAlignment.Start
     var verticalAlignment = CrossAxisAlignment.Start
+    var sizeListener: ResizableWidget.SizeListener? = null
 
     val children = UIViewChildren(
       container = this,
-      insert = { _, view, _, index ->
+      insert = { widget, view, _, index ->
+        if (widget is ResizableWidget<*>) {
+          widget.sizeListener = object : ResizableWidget.SizeListener {
+            override fun invalidateSize() {
+              this@View.invalidateSize()
+            }
+          }
+        }
         insertSubview(view, index.convert<NSInteger>())
-        view.setNeedsLayout()
       },
       remove = { index, count ->
         val views = Array(count) {
           typedSubviews[index].also(UIView::removeFromSuperview)
         }
-        setNeedsLayout()
         return@UIViewChildren views
       },
+      invalidateSize = ::invalidateSize,
     )
 
+    fun invalidateSize() {
+      val sizeListener = sizeListener
+      if (sizeListener != null) {
+        sizeListener.invalidateSize()
+      } else {
+        setNeedsLayout() // Update layout of subviews.
+      }
+    }
+
     override fun layoutSubviews() {
       super.layoutSubviews()
 
       children.widgets.forEach { widget ->
         val view = widget.value
+        view.setFrame(frame)
```

Comment:
This is a critical fix. Without this line the preceding size sticks.

Reply:
I assume this is where we need to apply margins and adjust each child's frame to fix #2003